nemo/tron/converter/common.py

```diff
@@ -0,0 +1,346 @@
+        del model
+
+    return state_dict
+
+
+def save_hf_tokenizer_assets(tokenizer_name_or_path, save_path="/tmp/nemo_tokenizer"):
+    """Save HF tokenizer to the imported NeMo model"""
+    from transformers import AutoTokenizer
+
+    tok = AutoTokenizer.from_pretrained(tokenizer_name_or_path, trust_remote_code=True)
+    # Save tokenizer assets to save_path.
+    tok.save_pretrained(save_path)
+    return save_path
+
+
+def dtype_from_str(dtype):
+    """
+    Convert a str precision to equivalent torch dtype.
+    """
+    assert isinstance(dtype, str)
+    if dtype in ["float16", "fp16", "16", "16-mixed"]:
+        return torch.float16
+    elif dtype in ["bfloat16", "bf16-mixed"]:
+        return torch.bfloat16
+    else:
+        return torch.float32
+
+
+def dtype_from_hf(config):
+    """
+    Extracts torch dtype from a HF config
+    """
+    assert hasattr(config, "torch_dtype"), "Expected config to have attr `torch_dtype`"
+    torch_dtype = config.torch_dtype
+    if isinstance(torch_dtype, torch.dtype):
+        return torch_dtype
+    elif isinstance(torch_dtype, str):
+        return dtype_from_str(torch_dtype)
+    else:
+        raise ValueError("torch_dtype is not of type str/torch.dtype")
+
+
+class _ModelState:
+    """
+    Helper class for used for to modify state dict of a source model during model conversion.
+    """
+
+    def __init__(self, state_dict):
+        self._state_dict = state_dict
+
+    def state_dict(self):
+        # pylint: disable=C0115,C0116
+        return self._state_dict
+
+    def to(self, dtype):
+        # pylint: disable=C0115,C0116
+        for k, v in self._state_dict.items():
+            if v.dtype != dtype:
+                logger.warning(f"Converting {k} from {v.dtype} (source model) to {dtype} (target model)")
+            self._state_dict[k] = v.to(dtype)
+
+
+class BaseImporter:
+    def __init__(self, input_path: str | Path, output_path: str | Path):
+        self.input_path = Path(input_path) if isinstance(input_path, str) else input_path
+        self.output_path = Path(output_path) if isinstance(output_path, str) else output_path
+        (self.output_path / HF_ASSETS_DIR).mkdir(parents=True, exist_ok=True)
+        self._hf_config = None
+        self._tron_config = None
+        self.hf_config.save_pretrained(str(self.output_path / HF_ASSETS_DIR))
+
+    @property
+    def tokenizer(self) -> "_HuggingFaceTokenizer":
+        """Get the tokenizer for the HF model.
+
+        Returns:
+            _HuggingFaceTokenizer: Tokenizer instance initialized from the HF model's tokenizer
+        """
+
+        return _HuggingFaceTokenizer(
+            save_hf_tokenizer_assets(str(self.input_path), str(self.output_path / HF_ASSETS_DIR))
+        )
+
+    def init_tron_model(self, cfg: GPTConfig | T5Config):
+        with megatron_cpu_init_context(cfg):
+            model = cfg.configure_model(tokenizer=self.tokenizer)
+        return [model]
```

Comment:
n00b question: why is this returned as a list of model parts? is this intended to be accessed anywhere outside `apply` ?

Reply:
The `save_checkpoint` method expects a list